File: Week01/9.回文数.py
#
# @lc app=leetcode.cn id=9 lang=python3
#
# [9] 回文数
#

# @lc code=start
class Solution:
    def isPalindrome(self, x: int) -> bool:
        # 暴力法：
        # 如果是负数，return False
        # 如果是正数，列表或字符串切片
        # str:80ms, 80.33%
        # list:100ms, 35.03%
        if x < 0: return False
        if len(str(x)) < 2: return True
        return str(x) == str(x)[::-1]
        # return list(str(x)) == list(str(x))[::-1]

s = Solution()
print(s.isPalindrome(-234))
print(s.isPalindrome(0))
print(s.isPalindrome(3))
print(s.isPalindrome(1328))
print(s.isPalindrome(3443))
# @lc code=end

File: Week01/9.回文数.java
/*
 * @lc app=leetcode.cn id=9 lang=java
 *
 * [9] 回文数
 */

// @lc code=start
class Solution {
    public boolean isPalindrome(int x) {
        // compare half of the digits in x, so don't need to deal with overflow.
        // 时间复杂度：O(N) 9ms 99.05%
        // 空间复杂度：O(1)
        // 负数，或者末位为0的非零数
        if (x<0 || (x!=0 && x%10==0)) return false;
        int rev = 0;
        while (x>rev) {
            rev = rev * 10 + x % 10;
            x = x/10;
        }
        return (x==rev || x==rev/10);
    }
}
// @lc code=end

